jvm/src/main/java/com/microsoft/bond/protocol/TaggedProtocolReader.java

```diff
@@ -0,0 +1,75 @@
+package com.microsoft.bond.protocol;
+
+import com.microsoft.bond.BondDataType;
+
+import java.math.BigInteger;
+
+/**
+ * Reads a serialized payload.
+ */
+public interface TaggedProtocolReader {
+    class ReadFieldResult {
+        public BondDataType type;
+        public int id;
+    }
+
+    class ReadContainerResult {
+        public long count;
+        public BondDataType element;
+        public BondDataType key;
```

Comment:
`null`. We could default to `BT_UNAVAILABLE` or whatever, but I like the fact that this code will throw a `NullPointerException` if you try to use one of these without passing it to someone.

The current implementation of `FastBinaryReader.readListBegin()` explicitly nulls `.key` for the same reason.